File: src/main/js/formulas/statistical/AVERAGEA.ts
import { isNotNull } from "../../utils/Other";
import { DivException } from "../../errors/DivException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Counter } from "../../utils/Counter";
import { Mappers } from "../../utils/Mappers";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class AVERAGEA extends AbstractFormula {
  static SELF: AVERAGEA = new AVERAGEA();
  NAME = FormulaName.AVERAGEA;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    const counter = new Counter<number>();
    const sum = values
      .map(this.lookup)
      .map(Mappers.flattenGridsToArrays) // 1, "2", [null, 1, "99", "Text.", null, #REF!, true]
      .map(Mappers.coerceNonArrayValuesToNumberOrThrow) // 1, 2, [null, 1, "99", "Text.", null, #REF!, true]
      .map(Mappers.ensureAllAreArrays) // [1], [2], [null, 1, "99", "Text.", null, #REF!, true]
      .reduce(Reducers.join) // [1, 2, null, 1, "99", "Text.", null, #REF!, true]
      .filter(isNotNull) // [1, 2, 1, "99", "Text.", #REF!, true]
      .map(Converters.textToZero) // [1, 2, 1, 0, 0, true]
      .map(Converters.toNumber) // [1, 2, 1, 0, 0, 1]
      .map((value) => counter.count(value))
      .reduce(Reducers.sum, 0);
    if (counter.getCount() === 0) {
      return new DivException(
        "Formula AVERAGEA caused an error when attempting to divide by zero."
      );
    }
    return sum / counter.getCount();
  }
}

File: src/main/java/io/protobase/f7/formulas/statistical/AVERAGEA.java
package io.protobase.f7.formulas.statistical;

import io.protobase.f7.errors.DivException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;
import io.protobase.f7.utils.Mappers;

import java.util.Arrays;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

public class AVERAGEA extends AbstractFormula {
  public static FormulaName NAME = FormulaName.AVERAGEA;
  public static AVERAGEA SELF = new AVERAGEA();

  public AVERAGEA() {
    super();
  }

  public AVERAGEA(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    List<Double> numbers = Arrays.stream(values)
        .map(lookup)
        .flatMap(Mappers::toFlatStreamFilterOnlyNumeric)
        .map(Converters::toDouble)
        .collect(Collectors.toList());
    if (numbers.size() == 0) {
      return new DivException("Formula AVERAGEA caused an error when attempting to divide by zero.");
    }
    double sum = 0;
    for (Double number : numbers) {
      sum += number;
    }
    return sum / numbers.size();
  }
}
